src/realm/util/sensitive_buffer.cpp

```diff
@@ -0,0 +1,230 @@
+        // Since we don't use VirtualAlloc with real page size and assign buffer to the portion
+        // of the allocated page (this'd complicate the logic significantly),
+        // simply try to double the limit a few times and attempt to lock the buffer again
+        // in case when multiple threads do the same. The min limit itself is pretty small,
+        // and the doc says that even this is not strictly guaranteed by the system, so
+        // there should be no harm in overcommiting expected min working set size
+        // In real tests: even with 128 threads on 16 cores system in 32bit app this loop
+        // may go through 4-5 iterations until the lock succeeds if the increment
+        // is just a multiple of a few pages, but with doubling of the limit
+        // it should succeed on the first try.
+        min_size *= 2;
+        max_size = std::max(4 * min_size, max_size);
+
+        // give up and don't try to ask the system to keep more than 90% memory resident in the process
+        if (max_size > 0.9 * limits->mem_size)
+            break;
+
+        BOOL ret = SetProcessWorkingSetSize(GetCurrentProcess(), min_size, max_size);
+        REALM_ASSERT_RELEASE_EX(ret != 0 && "SetProcessWorkingSetSizeEx", GetLastError());
+        limits->min_size = min_size;
+        limits->max_size = max_size;
+    }
+
+    // the loop above didn't succeed, provide some mem info and assert
+    MEMORYSTATUSEX mem;
+    mem.dwLength = sizeof(mem);
+    GlobalMemoryStatusEx(&mem);
+    REALM_ASSERT_RELEASE_EX(res != 0 && "VirtualLock()", err, mem.ullAvailPhys, mem.ullTotalPhys, mem.dwMemoryLoad,
+                            mem.ullAvailPageFile, mem.ullAvailVirtual);
+}
+
+SensitiveBufferBase::SensitiveBufferBase(size_t size)
+    : m_size(size_t(ceil((long double)size / CRYPTPROTECTMEMORY_BLOCK_SIZE) * CRYPTPROTECTMEMORY_BLOCK_SIZE))
+{
+    m_buffer = VirtualAlloc(nullptr, m_size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+    REALM_ASSERT_RELEASE_EX(m_buffer != NULL && "VirtualAlloc()", GetLastError());
+
+    lock_or_grow_working_size(m_buffer, m_size);
+}
+
+SensitiveBufferBase::~SensitiveBufferBase()
+{
+    if (!m_buffer)
+        return;
+
+    secure_erase(m_buffer, m_size);
+
+    BOOL ret = VirtualUnlock(m_buffer, m_size);
+    REALM_ASSERT_RELEASE_EX((ret != 0 || GetLastError() == ERROR_NOT_LOCKED) && "VirtualUnlock()", GetLastError());
+
+    ret = VirtualFree(m_buffer, 0, MEM_RELEASE);
+    REALM_ASSERT_RELEASE_EX(ret != 0 && "VirtualFree()", GetLastError());
+
+    m_buffer = nullptr;
+}
+
+void SensitiveBufferBase::protect() const
+{
+    // MEMO even though we try to lock the page with the buffer, and that should prevent it to be swapped,
+    //      locking is not reliable and may fail under high demand in the system due to some opaque reason,
+    //      which we can't recover from, so use second layer to protect the buffer if it is swapped
+    //      note: look at attempt_to_lock where it's expected
```

Comment:
This comment seems entirely backwards. CryptProtectMemory() is the actual "security" mechanism here and the thing we're guarding against is another process reading the key out of the application's memory. Stopping the page from getting swapped isn't useful without that.

Reply:
This is again me just appropriating mongod’s secure allocator wholesale without looking too deeply into it. Kirill and I discussed the likelihood that locking pages into RAM is superfluous if they’re encrypted in the first place.